bin/grade_item.py

```diff
@@ -0,0 +1,412 @@
+
+    # return to the main tmp directory
+    os.chdir(tmp)
+
+
+    # --------------------------------------------------------------------
+    # make the runner directory
+
+    with open(os.path.join(tmp_logs,"overall.txt"),'a') as f:
+        print ("====================================\nRUNNER STARTS", file=f)
+        
+    tmp_work=os.path.join(tmp,"TMP_WORK")
+    os.makedirs(tmp_work)
+    os.chdir(tmp_work)
+
+    # move all executable files from the compilation directory to the main tmp directory
+    # Note: Must preserve the directory structure of compiled files (esp for Java)
+
+    patterns_submission_to_runner = complete_config_obj["autograding"]["submission_to_runner"]
+    pattern_copy("submission_to_runner",patterns_submission_to_runner,submission_path,tmp_work,tmp_logs)
+
+    patterns_compilation_to_runner = complete_config_obj["autograding"]["compilation_to_runner"]
+    pattern_copy("compilation_to_runner",patterns_compilation_to_runner,tmp_compilation,tmp_work,tmp_logs)
+        
+    # copy input files to tmp_work directory
+    copy_contents_into(test_input_path,tmp_work)
+
+    subprocess.call("ls -la "+tmp_work+" >> "+tmp_logs+"/overall.txt",shell=True)
+    
+    # copy runner.out to the current directory
+    shutil.copy (os.path.join(bin_path,obj["gradeable"],"run.out"),os.path.join(tmp_work,"my_runner.out"))
+
+    # give the untrusted user read/write/execute permissions on the tmp directory & files
+    add_permissions_recursive(tmp_work,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH | stat.S_IXOTH)
+
+    # run the run.out as the untrusted user
+    with open(os.path.join(tmp_logs,"runner_log.txt"), 'w') as logfile:
+        runner_success = subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                                          args.which_untrusted,
+                                          os.path.join(tmp_work,"my_runner.out"),
+                                          obj["gradeable"],
+                                          obj["who"],
+                                          str(obj["version"]),
+                                          submission_time],
+                                          stdout=logfile)
+
+
+
+
+    if runner_success == 0:
+        print ("NEW RUNNER OK")
+    else:
+        print ("NEW RUNNER FAILURE")
+
+    # give permissions to all created files to the hwcron user
+    subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                     args.which_untrusted,
+                     "/usr/bin/find",
+                     tmp_work,
+                     "-user",
+                     args.which_untrusted,
+                     "-exec",
+                     "/bin/chmod",
+                     "o+r",
+                     "{}",
+                     ";"])
```

Comment:
Do we need this semi-colon? Looks like there's only one command here? Repeat for other versions of this one?